templates/pkg/resource/references.go.tpl

```diff
@@ -123,16 +138,44 @@ func resolveReferenceFor{{ $field.FieldPathWithUnderscore }}(
 	}
 	return nil
 }
+{{ else }}
+{{ $parentField := index .CRD.Fields $fp.String }}
+{{ if eq $parentField.ShapeRef.Shape.Type "list" -}}
```

Comment:
nit: I do not think you need this `if eq $parentField.ShapeRef.Shape.Type "list"` check(L#143) and it's corresponding else block(L#164).

Because after first `if` block ( `{{ if not $isList -}}` L#117)only lists will be passed down to `else if` and `else`. `else if`(L#127) checks for `nested lists` , and only `non-nested lists` will reach else block(L#141).

I don't think you need another shape check(L#143) inside `else` block.

Reply:
L#143 is not the same is `$isList` though. It checks that the parent of the field is a list. The first conditional is checking for single primitives, the second condition (the else if) is checking for lists of primitives. Within the third condition (the else) we are checking for lists of structs (the list is the parent field, not the field itself) and falling back to lists of primitives.

These are (some of) the conditions for resolving references - and which ones we support:
- [x] Single primitives (Condition 1)
- [x] List of primitives (Condition 2)
- [x] Singly-nested primitives (Condition 1)
- [x] Singly-nested struct (Condition 1)
- [x] Singly-nested list of primitives (Condition 2)
- [x] Singly-nested list of structs (Condition 3)
- [x] Recursively-nested primitives (Condition 1)
- [ ] Recursively-nested list of primitives
- [ ] Recursively-nested list of structs

Don't quote me on that. Perhaps some of these solutions can be generalised to match those, but it's also possible there are other cases I am not considering. Either way, I think this code will eventually need to be moved out of `references.go.tpl` and made with a string builder in Go. It's a bit too complex (especially if we use layers of iterators) for Go templates (and a PITA to debug).

Reply:
BTW I'd love to draw a [Karnaugh map](https://en.wikipedia.org/wiki/Karnaugh_map) of the conditions and figure out a minimised conditional for this